### DISCUSSION TITLE
Hajk 3.11 Release Candidate 1

### DISCUSSION THREAD

**Discussion by Hallbergs** [hajkmap/Hajk#1254] [General]
# Release candidate timeline
Alright, here's the latest RC! Please help out trying this out. 

**We usually run 3 weeks RC cycles, which would mean that you have until around Christmas holiday to report your findings!** 
/The Dev team, by @jacobwod 

---

_[236 commits](https://github.com/hajkmap/Hajk/compare/v3.10.0...master) and changes in 164 files since the 3.10 release – here's Hajk 3.11! This release was brought to you by @Hallbergs  and @jacobwod . Please see the list of commits for all contributors._

This release brings improvements across all across the solution, with a lot of efforts being put into **Print**, **Sketch**, the **Search** results list, **LayerSwitcher** and **Infoclick/MapClickViewer**, to name a few. 

There are of course upgrades to the latest dependencies that ensure that your Hajk setup is up-to-date when it comes to **the latest security patches**. 🔒  

Continue reading for a details list of changes! 🚀 

# Discussion and bug reports

Please use https://github.com/hajkmap/Hajk/discussions/ to report your findings from this release candidate.

# Changelog

## Breaking changes
None.

## (Almost) breaking changes
**NodeJS Backend only:** The NodeJS edition is now built using the current LTS version, namely Node 18. It brings a lot of new features and optimisations (such as native `Fetch` that [enables this nice layer verification](https://github.com/hajkmap/Hajk/issues/1227) in Backend). In order to use this new features, you must run the Backend on Node 18. 

**This is not a truly breaking change though: Hajk will still run on lower Node versions (tested on v14) – but you will not be able to use the new feature from #1227.**

### Improved group layer support for Infoclick/FeatureInfo/MapClickViewer
See #768.

### Secondary display fields in search results view (#1167)
Admins can now configure more informative search results lists by utilising the possibly to show more information for each result. The secondary label (or labels) will be shown below the main label, with a smaller font size: 

![image](https://user-images.githubusercontent.com/110222/205652295-b74aeaba-37d4-4194-bee5-389751488c73.png)

### Print: option to add a margin to the printout (#1174)
User can optionally add a margin to the PDF that will be created. There is another option to make the margin large enough for it to hold the norther arrow, title and logo too:
![image](https://user-images.githubusercontent.com/110222/205653631-3ce11b14-7fda-48a9-82e1-033acb39f53f.png)

### Sketch: possible to change draw order (z-index) of the drawn features (#1176)
Previously Sketch features were placed on map in an order that corresponded to the draw order (most recently drawn features were placed on top). 

This is still the default behaviour but this addition allows the user to control the z-index afterwards. This nice UI makes it possible to send features to front/back directly, or to move them one step at a time:
![image](https://user-images.githubusercontent.com/110222/205654303-02a74063-3e7e-4065-b7e5-ce62fdbceeaa.png)

### Allow users to change layer draw order (#1086)
A third tab, called `Active layers` can now be enabled in LayerSwitcher. This tab makes it possible for the user to rearrange the draw order (a.k.a. z-index) of the layers shown in map. It adds some other goodies too. 

Still a new feature that will be polished in future releases, but already a usable addition to our UI.
![image](https://user-images.githubusercontent.com/110222/205641963-859cc9e0-2df1-4448-acb5-5d99227c625a.png)

### MapClickViewer gains support for hard-coded values when no good attribute exists (#1126) 
Some layers need to be clickable but they may lack informative attribute values for features. This was problematic before, as we needed at least one parameter to create a nice label in MapClickViewer. 

With the release of 3.11, admins can hard-code a text string as an attribute value in Admin UI by surrounding the string with double quotations marks:

Admin UI:
![image](https://user-images.githubusercontent.com/110222/205643087-24e2e4ed-f2af-430f-bc0b-5fba4c1455ac.png)

Client UI:
![image](https://user-images.githubusercontent.com/110222/205642681-a9fbbd21-0b7f-4fab-8f4c-52a8c5390299.png)

### MarkDown and HTML are now allowed in Announcements (#1161)
We can now create more lively notifications using text formatting and even emojis!

![image](https://user-images.githubusercontent.com/110222/205651964-d1e92267-d538-47cc-8900-b1f5773d971c.png)

### Support for additional, non-latin alphabets in Print, Admin (#1181)
Thanks to the spread of Hajk outside of Sweden, we've discover – and fixed! – a bug related to displaying non-latin characters in some parts of the application. E.g. the PDFs created by Print can now include Cyrillic letters:
![image](https://user-images.githubusercontent.com/110222/205654872-04660022-205f-4647-a333-c22abdc21529.png)

### DocumentHandler window can now be resized and moved around (#1239)
Previously the DocumentHandler window was limited to being locked to the left side of the screen. With this addition, admins can opt-in to make the window behave just as any Hajk plugin.

### Web analytics expanded to also support Matomo (#1143)
Since #1065 we've had support for Plausible. In this release, we also add support for Matomo.

### Experimental: list of most recently used plugins (#1171)
This experimental feature adds a button to the bottom of the screen and makes it possible to easily toggle between most recently used plugins:
![image](https://user-images.githubusercontent.com/110222/205653248-756873d6-5a03-4c39-8488-3d06bd65ec49.png)

## New features and bug fixes
#883, #915, #1031, #1080, #1093, #1100, #1101, #1108, #1111, #1113, #1114, #1115, #1128, #1129, #1130, #1131, #1132, #1137, #1144, #1145, #1148, #1150, #1155, #1158, #1159, #1160, #1179, #1182, #1186, #1187, #1191, #1193, #1195, #1196, #1200, #1203, #1204, #1206, #1213, #1214, #1215, #1216, #1223, #1224, #1225, #1227, #1231, #1233, #1234, #1245, #1248

# Downloads
There are multiple ways to get the latest Hajk up and running, including Docker container.

## Pre-compiled releases
As usual, we have prepared **three different** versions of the pre-compiled solution (scroll down for download links):
1. [Hajk 3.11 NodeJS edition](https://github.com/hajkmap/Hajk/releases/download/v3.11.0-rc.1/hajk-v3.10.0-rc.1-nodejs.zip)
1. [Hajk 3.11 Simple edition](https://github.com/hajkmap/Hajk/releases/download/v3.11.0-rc.1/hajk-v3.10.0-rc.1-simple.zip)
1. [Hajk 3.11 .NET edition](https://github.com/hajkmap/Hajk/releases/download/v3.11.0-rc.1/hajk-v3.10.0-rc.1-dotnet.zip)

## Docker container
The container is available on Docker Hub: https://hub.docker.com/r/hallbergs/hajk

# Installation
Refer to our [installation guide](https://github.com/hajkmap/Hajk/wiki/Installation-guide-%28for-pre-packaged-releases%29).

------

## What's Changed
* Feature/883 layer naming by @jacobwod in https://github.com/hajkmap/Hajk/pull/1094
* Allow users to see active layers, control draw order, transparency/toggeling as well as save and restore functionality by @jacobwod in https://github.com/hajkmap/Hajk/pull/1116
* Bump Newtonsoft.Json from 5.0.4 to 13.0.1 in /proxy/mvc by @dependabot in https://github.com/hajkmap/Hajk/pull/1120
* Bump Newtonsoft.Json from 5.0.4 to 13.0.1 in /proxy/HTTPProxy/Proxy by @dependabot in https://github.com/hajkmap/Hajk/pull/1121
* Bump Newtonsoft.Json from 12.0.3 to 13.0.1 in /backend/mapservice by @dependabot in https://github.com/hajkmap/Hajk/pull/1122
* Bump Newtonsoft.Json from 10.0.3 to 13.0.1 in /backend/EdpConn by @dependabot in https://github.com/hajkmap/Hajk/pull/1123
* Bump Newtonsoft.Json from 5.0.4 to 13.0.1 in /proxy/GeoServerProxy/Proxy by @dependabot in https://github.com/hajkmap/Hajk/pull/1124
* Bump moment from 2.29.1 to 2.29.4 in /new-admin by @dependabot in https://github.com/hajkmap/Hajk/pull/1133
* Bump terser from 4.8.0 to 4.8.1 in /new-admin by @dependabot in https://github.com/hajkmap/Hajk/pull/1135
* Bump terser from 5.12.1 to 5.14.2 in /new-client by @dependabot in https://github.com/hajkmap/Hajk/pull/1136
* Add options and MUI5 upgrades in Geosuite plugin by @linusfj in https://github.com/hajkmap/Hajk/pull/1134
* Feature/1101 information to activate location by @linusfj in https://github.com/hajkmap/Hajk/pull/1119
* Added possiblity for the user to set a custom print-scale. by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1140
* Feature/915 custom hdpi by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1141
* Upgrade UI for Coordinates plugin by @linusfj in https://github.com/hajkmap/Hajk/pull/1138
* Feature/1150 infoclick multiply by by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1151
* Bump async and async in /new-client by @dependabot in https://github.com/hajkmap/Hajk/pull/1142
* Bookmark Removal Confirmation by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1157
* Fix/1115 admin wms projection by @linusfj in https://github.com/hajkmap/Hajk/pull/1164
* Feature/1160 sketch replace warning by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1166
* feature/1143 analytics matomo by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1168
* Feature/1145 infoclick xy zoom by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1163
* Added support to mimic proxyBaseUrl in GeoServer by @maan002 in https://github.com/hajkmap/Hajk/pull/1170
* Fix automatic zoom on mobile devices by @linusfj in https://github.com/hajkmap/Hajk/pull/1169
* Feature/1093 new dummy plugin by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1173
* Added a plugin-template by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1175
* Feature/1171 plugin history by @jacobwod in https://github.com/hajkmap/Hajk/pull/1180
* Feature/1174 print marginal by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1183
* Fix/1182 single layer in group qgis server by @jacobwod in https://github.com/hajkmap/Hajk/pull/1188
* Fix/1181 cyrillic print support by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1189
* Potential fix for #1186 by @Hallbergs in https://github.com/hajkmap/Hajk/pull/1190
* Feature/1191 placement warning by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1192
* Fix/1195 positioning bug by @linusfj in https://github.com/hajkmap/Hajk/pull/1197
* Added date prop to print by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1199
* Control buttons render in Drawer on small screens, closes #1193. by @jacobwod in https://github.com/hajkmap/Hajk/pull/1194
* Bump @xmldom/xmldom from 0.7.5 to 0.7.6 in /new-client by @dependabot in https://github.com/hajkmap/Hajk/pull/1218
* Bump @xmldom/xmldom from 0.7.5 to 0.7.6 in /new-admin by @dependabot in https://github.com/hajkmap/Hajk/pull/1217
* Fix/1204 infoclick title by @linusfj in https://github.com/hajkmap/Hajk/pull/1205
* Feature/fir update 1 by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1235
* Feature/1239 documenthandler resize dragging by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1240
* Feature/snackbar close button by @linusfj in https://github.com/hajkmap/Hajk/pull/1232
* Feature/1227 verify layers by @jacobwod in https://github.com/hajkmap/Hajk/pull/1242
* Feature/786 grouplayer infoclick by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1241
* Added group layers support to Anchor by @jacobwod in https://github.com/hajkmap/Hajk/pull/1246
* Feature/orebro tsm by @Sunix71 in https://github.com/hajkmap/Hajk/pull/1228
* Feature/1176 sketch control zindex by @jesade-vbg in https://github.com/hajkmap/Hajk/pull/1249
* 1200 admin marginals and date by @OlofSvahnVbg in https://github.com/hajkmap/Hajk/pull/1201

## New Contributors
* @OlofSvahnVbg made their first contribution in https://github.com/hajkmap/Hajk/pull/1157
* @Sunix71 made their first contribution in https://github.com/hajkmap/Hajk/pull/1228

**Full Changelog**: https://github.com/hajkmap/Hajk/compare/v3.10.0...v3.11.0-rc.1

<hr /><em>This discussion was created from the release <a href='https://github.com/hajkmap/Hajk/releases/tag/v3.11.0-rc.1'>Hajk 3.11 Release Candidate 1</a>.</em>

**Comment by bitmapbulgaria** [comment]
Awesome work, guys! 
Is this version have Multilanguage support?

**Reply by jacobwod** [reply]
No, that job is ongoing in the `feature/502-internationalization` branch but it's not done yet. Help is welcome! 😄 

(This release does include some related fixes though, such as #1189.)

**Comment by elizabethgbg** [comment]
Nice work! We've started testing and this is what we've found so far. We’ll let you know if we find anything more.

#1239
Works very well!
Suggestions for further development: a possibility to manage “window settings” (fönsterinställningar in Swedish admin) for document manager 2.0 in admin, as you can do for several other tools in admin (eg the draw tool). The possibility to manage the window placement could be a useful funktion, for example.

#1176
The function works very well! We think that it would be good if the ”explanatory text” in the tool, see circled in the image below, also describes these new functions - to alert the user of these functions.
![Skärmklipp](https://user-images.githubusercontent.com/120728926/208125409-1811e9ff-0918-4eab-86d6-3de358ecb7cd.PNG)

**Reply by Hallbergs** [reply]
> Nice work! We've started testing and this is what we've found so far. We’ll let you know if we find anything more.
> 
> #1239 Works very well! Suggestions for further development: a possibility to manage “window settings” (fönsterinställningar in Swedish admin) for document manager 2.0 in admin, as you can do for several other tools in admin (eg the draw tool). The possibility to manage the window placement could be a useful funktion, for example.
> 
> #1176 The function works very well! We think that it would be good if the ”explanatory text” in the tool, see circled in the image below, also describes these new functions - to alert the user of these functions. ![Skärmklipp](https://user-images.githubusercontent.com/120728926/208125409-1811e9ff-0918-4eab-86d6-3de358ecb7cd.PNG)

Just fixed the texts for #1176! Something like this should do?
![image](https://user-images.githubusercontent.com/43059093/208358182-319aa79d-871c-4e59-af8b-355726449b3d.png)

**Reply by elizabethgbg** [reply]
Yes that's perfect!

**Comment by ingvar-uddevalla** [comment]
I don't know if this is an on-going issue, because it hasn't worked in previous versions either ... or maybe I'm doing it wrong. 
It's about Infoclick and URLs. If admin uses "standard"-format for Infoclick (leave "Inforuta" empty) the user will not be able to open an URL-adress. See example below.
![Skärmbild 2022-12-19 120659](https://user-images.githubusercontent.com/54808837/208413915-3c1e426e-d972-4cdd-9e54-09dcb1d8be8f.png)
To make it work the administrator has to define the attribute-fields in "Inforuta". 
![image](https://user-images.githubusercontent.com/54808837/208414296-3d5bfec4-f794-48d4-a4e0-f6fd18ea5e5b.png)
And then it's possible to open the site when clicking on the address.
![image](https://user-images.githubusercontent.com/54808837/208414996-fb80aefa-b689-40c0-81bb-6d64c8ed3731.png)

**Reply by Hallbergs** [reply]
The default attribute table just spits out all attributes and their values - working as intended. I don't think we've talked about adding support for highlighted url:s in this release. 

I would suggest that you create an issue instead.

**Comment by elizabethgbg** [comment]
We’ve now competed our tests. We’ve tested a large part of the new features and bug fixes. These are our remaining findings:

#1186
Works very well! However, we think it would be good if the colors black and white are selectable in the list of colors, so that you don't need the color codes to be able to switch to these colors.

#1196
The function works well! A tooltip would be great in admin to inform the admin of how to set this up.

#1203
Which settings must be made in admin to make the size of the north arrow adapt to the map image? We can't quite make it work. The size of the map image and the north arrow don’t seem to be affected by each other in our tests, see images below.
![Skärmklipp](https://user-images.githubusercontent.com/120728926/208615326-03848968-5a7c-4012-a57b-c30fb50c0f9d.PNG)

**Comment by elizabethgbg** [comment]
Hi! One more finding/question:

We would like to make changes to this button's settings and can't seem to find where we can do that, see button in the picture below. Is it possible to change the settings?
![Skärmklipp](https://user-images.githubusercontent.com/120728926/208705542-8f6e1edb-a2b4-46dc-b5ff-26bf3e84f341.PNG)

It seems like the “Document”-button in the drawer needs a specified "startdokument" to work – if a startdokument isn’t specified we get a loading window if we push the document-button in the drawer, see picture below.
![Skärmklipp1](https://user-images.githubusercontent.com/120728926/208705560-741e8afc-2238-4ae7-925c-20b36904ceb0.PNG)
We also can’t seem to find where to remove the document-button from the drawer.

Settings we're searching for:
- The possibility to choose whether to use the button or not. (At the moment we can't find where to remove the button from the drawer).
- The possibility to change the name of the button.
- The possibility to use the button without a specified startdokument. (At the moment we only get a loading window when we push the button if a startdokument isn’t specified).

**Reply by Hallbergs** [reply]
It was a bug causing the document handler button to show in the toolbar. It shouldn't have been there at all. The bug is fixed now and the result will show in rc2.

**Comment by Hallbergs** [comment]
#1263 fixed and added to 3.11. Will be visible in rc2.

**Comment by Hallbergs** [comment]
Fixed a bug where properties from drawn features (from the sketch plugin etc.) were trying to render in the info-click.

**Comment by elizabethgbg** [comment]
Hi @Hallbergs!
These findings, se below, do you know if there are any plans for them to be handled in future development or should perhaps new issues be created for this? 

[#1239](https://github.com/hajkmap/Hajk/issues/1239)
Works very well!
Suggestions for further development: a possibility to manage “window settings” (fönsterinställningar in Swedish admin) for document manager 2.0 in admin, as you can do for several other tools in admin (eg the draw tool). The possibility to manage the window placement could be a useful funktion, for example.

[#1186](https://github.com/hajkmap/Hajk/issues/1186)
Works very well! However, we think it would be good if the colors black and white are selectable in the list of colors, so that you don't need the color codes to be able to switch to these colors.

[#1196](https://github.com/hajkmap/Hajk/issues/1196)
The function works well! A tooltip would be great in admin to inform the admin of how to set this up.

[#1203](https://github.com/hajkmap/Hajk/issues/1203)
Which settings must be made in admin to make the size of the north arrow adapt to the map image? We can't quite make it work. The size of the map image and the north arrow don’t seem to be affected by each other in our tests, see images below.

**Reply by Hallbergs** [reply]
Hi! Sorry, i just saw that i've missed to respond to those. I think they should be added as new issues. Regarding the last question, maybe @OlofSvahnVbg knows the answer?
